TELCODOCS-1827: Document symlinks for in-tree dependencies (#76114)
openshift/openshift-docs: modules/kmm-symbolic-links-for-in-tree-dependencies.adoc

// Module included in the following assemblies:
//
// * hardware_enablement/kmm-kernel-module-management.adoc

:_mod-docs-content-type: CONCEPT
[id="kmm-symbolic-links-for-in-tree-dependencies_{context}"]

= Symbolic links for in-tree dependencies

> Some kernel modules depend on other kernel modules that are shipped with the node's distribution.

Reviewer: +1

> To avoid copying those dependencies into the kmod image, KMM mounts `/usr/lib/modules` into both the build and the

Reviewer 2: I think that we should define KMM in this first usage. 

```suggestion
To avoid copying those dependencies into the kmod image, Kernel Module Management  (KMM) mounts `/usr/lib/modules` into both the build and the
```

Author: added.

worker pod's filesystems.

> By creating a symlink from `/opt/usr/lib/modules/[kernel-version]/[symlink-name]` to

Reviewer 2: Conflicted as to whether these count as user-replaced values. Based on https://redhat-documentation.github.io/supplementary-style-guide/#user-replaced-values, I think that they are (User-replaced values are often found in places such as code blocks, file paths, and commands.). If we're in agreement that they are, they should be written as follows:

```suggestion
By creating a symlink from `/opt/usr/lib/modules/<kernel_version>/<symlink_name>` to
```

Author: Changed. FYI - The bracket style used is the nomenclature for the mid- and upstream versions. But, these are indeed user-replaced values. I'll make all changes.

> `/usr/lib/modules/[kernel-version]`, `depmod` can use the in-tree kmods on the building node's filesystem to resolve

Reviewer 2: Same sentiment here.

```suggestion
`/usr/lib/modules/<kernel_version>`, `depmod` can use the in-tree kmods on the building node's filesystem to resolve
```

Author: changed.

dependencies.

> At runtime, the worker pod extracts the entire image, including the `[symlink-name]` symbolic link.

Reviewer 2: Same sentiments with this one, but maybe it's different given the context? I'm going to assume that it's a user-replaced variable here as well:

```suggestion
At runtime, the worker pod extracts the entire image, including the `<symlink_name>` symbolic link.
```

Author: changed.

> That link points to `/usr/lib/modules/[kernel-version]` in the worker Pod, which is mounted from the node's filesystem.

Reviewer 2: Slight suggestion. Also same sentiments as above. This is the last time that I'll point it out if you opt to change based on the RHSSG. 
```suggestion
That symbolic link points to `/usr/lib/modules/<kernel_version>` in the worker pod, which is mounted from the node's filesystem.
```

Author: Made all changes.

`modprobe` can then follow that link and load the in-tree dependencies as needed.

In the following example, `host` is the symbolic link name under `/opt/usr/lib/modules/[kernel-version]`:

[source,dockerfile]
----
ARG DTK_AUTO

FROM ${DTK_AUTO} as builder

#
# Build steps
#

FROM ubi9/ubi

ARG KERNEL_FULL_VERSION

RUN dnf update && dnf install -y kmod

COPY --from=builder /usr/src/kernel-module-management/ci/kmm-kmod/kmm_ci_a.ko /opt/lib/modules/${KERNEL_FULL_VERSION}/
COPY --from=builder /usr/src/kernel-module-management/ci/kmm-kmod/kmm_ci_b.ko /opt/lib/modules/${KERNEL_FULL_VERSION}/

# Create the symbolic link
RUN ln -s /lib/modules/${KERNEL_FULL_VERSION} /opt/lib/modules/${KERNEL_FULL_VERSION}/host

RUN depmod -b /opt ${KERNEL_FULL_VERSION}
----

[NOTE]
====

> `depmod` will generate dependency files based on the kernel modules present on the node that runs the kmod

Reviewer 2: Suggestion to remove "will" based on our guidelines to avoid future tense. https://redhat-documentation.github.io/supplementary-style-guide/#will

```suggestion
`depmod` generates dependency files based on the kernel modules present on the node that runs the kmod
```

Author: changed.

image build.

> On the node on which KMM loads the kernel modules, `modprobe` will expect the files to be present under

Reviewer 2: Same here:

```suggestion
On the node on which KMM loads the kernel modules, `modprobe` expects the files to be present under
```

Author: changed.

Author: @stevsmit I also fixed the line wrapping. :)

`/usr/lib/modules/[kernel-version]`, and the same filesystem layout.

> It is highly recommended that the build and the target nodes share the same distribution and release.

Reviewer 3: ```suggestion
It is highly recommended that the build and the target nodes share the same operating system and release.
```

Reviewer: +1

Author: changed.

> It is highly recommended that the build and the target nodes share the same distribution and release.

Reviewer 3: ```suggestion
It is highly recommended that the build and the target nodes share the same operating system and release.
```

Reviewer: +1

Author: changed.
